### PR TITLE
Add `JsNumberToInt64` function

### PR DESCRIPTION
**Pull request by kfarnung**
Allow the caller to get the integer value for a given number back as a
64-bit integer value.

### CODE REVIEW

**Comment by kfarnung** [comment]
I had written this up when trying to fix a bug in node-chakracore, but ended up just using a static cast from double to int64_t instead.  Looking to see if other people think this would be useful.

File: lib/Jsrt/Jsrt.cpp [hunk 1]
```diff
@@ -1157,6 +1157,31 @@ CHAKRA_API JsNumberToInt(_In_ JsValueRef value, _Out_ int *asInt)
     END_JSRT_NO_EXCEPTION
 }
 
+CHAKRA_API JsNumberToInt64(_In_ JsValueRef value, _Out_ int64_t *result)
+{
+    return ContextAPIWrapper<JSRT_MAYBE_TRUE>([&] (Js::ScriptContext *scriptContext, TTDRecorder& _actionEntryPopper) -> JsErrorCode {
+        PERFORM_JSRT_TTD_RECORD_ACTION_NOT_IMPLEMENTED(scriptContext);
```
Comment (kfarnung) [inline lib/Jsrt/Jsrt.cpp]: I should fix this before landing this PR

**Comment by MSLaguana** [comment]
Is this only guaranteed to be precise for 52 bit numbers?

**Comment by kfarnung** [comment]
It's still subject to the significant bit limits of the IEEE754 double-precision numbers.  In this case it's 53-bits of integer precision (technically 54 if you count the sign bit).

I can add that remark to the documentation if there's agreement that this would be useful.

**Comment by kfarnung** [comment]
@MSLaguana I updated the comment to call out the precision caveat.

**Comment by kfarnung** [comment]
@boingoing Can you take another look?

File: lib/Jsrt/Jsrt.cpp [hunk 2]
```diff
@@ -1157,6 +1157,30 @@ CHAKRA_API JsNumberToInt(_In_ JsValueRef value, _Out_ int *asInt)
...
+
+    BEGIN_JSRT_NO_EXCEPTION
+    {
+        if (Js::TaggedInt::Is(value))
+        {
+            *result = Js::TaggedInt::ToInt64(value);
+        }
+        else if (Js::JavascriptNumber::Is_NoTaggedIntCheck(value))
+        {
+            *result = Js::JavascriptConversion::ToInt64(Js::JavascriptNumber::GetValue(value));
```
Comment (MSLaguana) [inline lib/Jsrt/Jsrt.cpp]: Does this case apply to `Number` objects, e.g. the result of `new Number(1)`? If so, this conversion can be re-entrant into script via a custom `valueOf` I think?

File: lib/Runtime/Language/JavascriptConversion.cpp [hunk 3]
```diff
@@ -1256,6 +1256,11 @@ namespace Js
         return JavascriptMath::ToInt32Core(T1);
     }
 
+    __int64 JavascriptConversion::ToInt64(double T1)
+    {
+        return JavascriptMath::TryToInt64(T1);
```
Comment (boingoing) [inline lib/Runtime/Language/JavascriptConversion.cpp]: What should we do if `NumberUtilities::IsValidTryToInt64` is false for the result of this call?

File: lib/Jsrt/Jsrt.cpp (hunk 2 above)
Comment (digitalinfinity) [reply lib/Jsrt/Jsrt.cpp]: No, this is explicitly checking for the tagged number case/JavascriptNumber case. Number objects are JavascriptNumberObjects. That being said, should we be checking for JavascriptNumberObject here? It looks like the existing implementations of JsNumberTo* don't handle Number objects, they likely require the caller to call JsConvertValueToNumber first.

File: lib/Runtime/Language/JavascriptConversion.cpp (hunk 3 above)
Comment (digitalinfinity) [reply lib/Runtime/Language/JavascriptConversion.cpp]: Seems like TryToInt64 would handle this case? In the case of overflows/Nans, it would just return canonical values? We should however document it

File: bin/NativeTests/JsRTApiTest.cpp [hunk 4]
```diff
@@ -521,15 +521,38 @@ namespace JsRTApiTest
...
+        REQUIRE(JsDoubleToNumber(9007199254740991, &value) == JsNoError);
+        REQUIRE(JsNumberToInt(value, &intValue) == JsNoError);
+        CHECK(-1 == intValue);
+        REQUIRE(JsNumberToInt64(value, &int64Value) == JsNoError);
+        CHECK(9007199254740991 == int64Value);
+
+        // Number.MIN_SAFE_INTEGER
+        REQUIRE(JsDoubleToNumber(-9007199254740991, &value) == JsNoError);
+        REQUIRE(JsNumberToInt(value, &intValue) == JsNoError);
+        CHECK(1 == intValue);
```
Comment (digitalinfinity) [inline bin/NativeTests/JsRTApiTest.cpp]: Mild preference for defining Pos_InvalidInt64/Neg_InvalidInt64 as JSRT constants, and compare against that?

File: bin/NativeTests/JsRTApiTest.cpp [hunk 5]
```diff
@@ -521,15 +521,38 @@ namespace JsRTApiTest
...
+        REQUIRE(JsNumberToInt64(value, &int64Value) == JsNoError);
+        CHECK(-9007199254740991ll == int64Value);
```
Comment (digitalinfinity) [inline bin/NativeTests/JsRTApiTest.cpp]: Can you add some test cases for Nan and overflow too?

File: lib/Jsrt/ChakraCore.h [hunk 6]
```diff
@@ -1063,5 +1063,22 @@ CHAKRA_API
         _Out_opt_ JsValueRef* target,
         _Out_opt_ JsValueRef* handler);
 
+///     Retrieves the <c>int64_t</c> value of a number value.
+/// </summary>
+/// <remarks>
```
Comment (irinayat-MS) [inline lib/Jsrt/ChakraCore.h]: Nit: missing opening <summary> tag

Comment (kfarnung) [reply lib/Jsrt/ChakraCore.h]: Thanks!  I must have dropped it when I was rebasing.

File: bin/NativeTests/JsRTApiTest.cpp (hunk 4 above)
Comment (kfarnung) [reply bin/NativeTests/JsRTApiTest.cpp]: In this case it's not actually invalid, it's the result of overflowing the int32, I'm still trying to figure out what to do for truly invalid numbers that exceed the bounds of int64, those end up returning `0x8000000000000000` on x86/x64, but that doesn't seem to be 100% consistent across architectures.

File: lib/Runtime/Language/JavascriptConversion.cpp (hunk 3 above)
Comment (kfarnung) [reply lib/Runtime/Language/JavascriptConversion.cpp]: `TryToInt64` doesn't do anything special for x64 builds, it's simply a cast from `double` to `int64_t`.  There are special paths for x86 (which can use SSE3) and ARM (which has different behavior), but in the base case it's just a cast.

For `ToInt32` it ends up normalizing all of these cases down to zero without returning an error, but I'm not sure if that's spec'd anywhere.

**Comment by liminzhu** [comment]
Just saw this PR. Btw there's a new primitive [BigInt](https://github.com/tc39/proposal-bigint) proposal in TC39 (stage 3 already, pretty close to standardized), which allows representing int64 in JS. It would complicate things a bit if this API also takes BigInt.

**Comment by kfarnung** [comment]
There are some open questions about how this API should handle numbers too large to represent even without considering BigInt.  For Int32 we seem to just wrap back around, but for Int64 the default behavior is to return a number indicating the invalid state (some CPUs do this by returning 0x8000000000000000 and we appear to use that ourselves).

I suspect the same sort of issues will occur with BigInt, except that 0x8000000000000000 will be possible to represent exactly with a BitInt when it's not possible to do so using a double.

Any thoughts about what the API should do in cases where the number can't be represented in an Int64 value?

**Comment by liminzhu** [comment]
> for Int64 the default behavior is to return a number indicating the invalid state 

It would be good to study some prior art. Can you point me to what language/platform use this as a default behavior? I was thinking more along the lines of returning a overflow error code, since you'd be getting that from .NET if you try to convert a bigint too large to int64.

I looked a bit more into BigInt and it looks like the proposal champion wants to set a clear boundary between BigInt and Number by having them as separate primitives and disallowing implicit conversions and ops mix-up, which is an approach I can agree with. This also makes me feel that if we were to expose BigInt natively, we want a different API, and b/c there is no builtin in C corresponding directly to bigint, we'll have to think about that later.

**Comment by alexgwolff** [comment]
hey, what is the final decision on this issue? I need to convert a Int64 in .net environment.
thanks

**Comment by alexgwolff** [comment]
@kfarnung  ??

**Comment by kfarnung** [comment]
@alexgwolff technically you can just convert the number to a double and then cast to a 64-bit integer.  Since JavaScript numbers are all IEEE754, that's essentially what this code does internally.

The problem was that we couldn't find agreement on how to represent invalid conversions, that behavior isn't defined by the standard and V8 seemed to use the value that the base conversion returns (`0x8000000000000000`), but that's basically an architecture-specific magic number.
